### PR TITLE
Allow to specify a Docker secret as apprepo credentials

### PR DESCRIPTION
**Pull request by andresmgot**
### Description of the change

Allow to specify a secret with Docker registry credentials in the AppRepository spec. For doing so it's necessary to modify the spec and add the logic to retrieve the credentials from the Docker registry secret on its format.

### Benefits

No need to duplicate auth info.

### Applicable issues

  - ref #2425

### CODE REVIEW

File: cmd/apprepository-controller/pkg/apis/apprepository/v1alpha1/types.go [hunk 1]
```diff
@@ -56,18 +56,13 @@ type AppRepositorySpec struct {
 
 // AppRepositoryAuth is the auth for an AppRepository resource
 type AppRepositoryAuth struct {
-	Header   *AppRepositoryAuthHeader `json:"header,omitempty"`
-	CustomCA *AppRepositoryCustomCA   `json:"customCA,omitempty"`
+	Header        *AppRepoAuthSecret `json:"header,omitempty"`
+	CustomCA      *AppRepoAuthSecret `json:"customCA,omitempty"`
+	RegistryCreds *AppRepoAuthSecret `json:"registryCreds,omitEmpty"`
```
Comment (antgamdia) [inline cmd/apprepository-controller/pkg/apis/apprepository/v1alpha1/types.go:61]: This change will also imply adding the corresponding field in  the API description :sweat_smile::
https://github.com/kubeapps/kubeapps/blob/master/dashboard/public/openapi.yaml#L2104-L2111

File: pkg/chart/chart.go [hunk 2]
```diff
@@ -394,6 +404,13 @@ func (c *OCIClient) InitClient(appRepo *appRepov1.AppRepository, caCertSecret *c
 			return err
 		}
 		headers.Set("Authorization", string(auth))
+	} else if registryCreds != nil && appRepo.Spec.Auth.RegistryCreds != nil {
```
Comment (antgamdia) [inline pkg/chart/chart.go:407]: Just to be sure: there is no case in which both `authSecret` and `registryCreds` are being passed, no? If so, we should say somewhere that the latter takes precedence over the former (and add a test with both parameters).
If not, this ok as is.

File: pkg/kube/http_utils.go [hunk 3]
```diff
@@ -71,6 +74,38 @@ func GetData(key string, s *corev1.Secret) (string, error) {
 	return auth, nil
 }
 
+// GetDataFromRegistrySecret retrieves the given key from the secret as a string
+func GetDataFromRegistrySecret(key string, s *corev1.Secret) (string, error) {
+	dockerConfigJsonEncoded, ok := s.StringData[key]
```
Comment (antgamdia) [inline pkg/kube/http_utils.go:79]: Will this always be a _Docker_ registry? Shouldn't we be as repo-agnostic as possible?
Edit: I see there is a `Type:  kubernetes.io/dockerconfigjson`, so that's this function is meant for.  Are you aware of other registry credentials rather than that one?

File: pkg/kube/kube_handler.go [hunk 4]
```diff
@@ -567,7 +568,7 @@ func (a *userHandler) DeleteAppRepository(repoName, repoNamespace string) error
 	return err
 }
 
-func getValidationCli(appRepoBody io.ReadCloser, requestNamespace, kubeappsNamespace string) (*v1alpha1.AppRepository, HTTPClient, error) {
+func (a *userHandler) getValidationCli(appRepoBody io.ReadCloser, requestNamespace, kubeappsNamespace string) (*v1alpha1.AppRepository, HTTPClient, error) {
```
Comment (antgamdia) [inline pkg/kube/kube_handler.go:571]: Out of curiosity: why this change?
Edit: ah, for the test!

**Review by antgamdia — APPROVED** [review]
Great! It will simplify the UX when adding a new repo. Thank you for the changes!

File: pkg/kube/kube_handler.go (hunk 4 above)
Comment (andresmgot) [reply pkg/kube/kube_handler.go:571]: this is executed when running the apprepo validation. We need now to fetch the secret from the namespace so we need to use the user clientset (so we can retrieve the credentials).

File: pkg/chart/chart.go (hunk 2 above)
Comment (andresmgot) [reply pkg/chart/chart.go]: I have changed this to use the same secret and treat the case in which the secret key is `.dockerconfigjson` differently.
